condor_git_config.py

```diff
@@ -91,127 +98,174 @@ class ConfigCache(object):
+        self._config_meta: Optional[Dict[str, Any]] = None
 
-    def abspath(self, *rel_paths):
-        return os.path.abspath(os.path.join(self._work_path, *rel_paths))
+    @property
+    def stats(self):
+        """Statistics of the cache"""
+        meta = self._config_meta if self._config_meta is not None else self._read_meta()
+        age = time.time() - meta["timestamp"]
+        return {"age": age, "pulls": meta["pulls"], "reads": meta["reads"]}
 
-    def repo_path(self, *rel_paths):
-        return self.abspath('repo', *rel_paths)
+    def _read_meta(self):
+        try:
+            with open(self._meta_file, "r") as meta_stream:
+                meta_data = json.load(meta_stream)
+        except FileNotFoundError:
+            return {
+                "git_uri": self.git_uri,
+                "branch": self.branch,
+                "timestamp": 0.0,
+                "pulls": 0,
+                "reads": 0,
+            }
+        else:
+            if meta_data["git_uri"] != self.git_uri:
+                LOGGER.critical("cache %r corrupted by other hook: %r", self, meta_data)
+                raise RuntimeError(
+                    "config cache %r used for conflicting hooks" % self.cache_path
+                )
+            return meta_data
+
+    def _write_meta(self):
+        if self._config_meta is None:
+            return
+        with open(self._meta_file, "w") as meta_stream:
+            json.dump(self._config_meta, meta_stream)
+
+    def abspath(self, *rel_paths: Union[str, Path]) -> Path:
+        return self._work_path.joinpath(*rel_paths)
+
+    def repo_path(self, *rel_paths: Union[str, Path]) -> Path:
+        return self.abspath("repo", *rel_paths)
 
     def __enter__(self):
         self._cache_lock.acquire()
+        self._config_meta = self._read_meta()
+        self._refresh()
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
+        self._write_meta()
+        self._config_meta = None
         self._cache_lock.release()
         return False
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable[Path]:
+        assert self._config_meta is not None
         # avoid duplicates from links
         seen = set()
         repo_path = self.repo_path()
         for dir_path, dir_names, file_names in os.walk(repo_path):
-            if '.git' in dir_names:
-                dir_names.remove('.git')
-            dir_names[:] = sorted(dir_names)
+            try:
+                dir_names.remove(".git")
+            except ValueError:
+                pass
+            dir_names.sort()
+            dir_path = Path(dir_path)
             for file_name in sorted(file_names):
-                rel_path = os.path.normpath(os.path.relpath(os.path.join(dir_path, file_name), repo_path))
+                rel_path = (dir_path / file_name).relative_to(repo_path)
                 if rel_path in seen:
                     continue
                 seen.add(rel_path)
```

Comment:
I am not sure if you actually need to check for duplicates here. Would have to test that. Do you have something specific in mind?

Reply:
No, actually I don't. I guess it's better to remove it until there is a practical case that needs it.

Reply:
I've removed the duplicate filter.